client/shared/src/search/query/patternMatcher.ts

```diff
@@ -0,0 +1,408 @@
+ * for convenience. See the combinatorial helper functions below for examples.
+ */
+export type PatternFunction<Value, Data> = (
+    value: Value,
+    context: MatchContext<Data>,
+    matchfn: typeof matches
+) => boolean
+
+/**
+ * This type is used for matching against primitives. The main reason for it's
+ * existence is to allow first class support matching strings with regular
+ * expressions.
+ */
+type PrimitivePattern<Value> = Value extends string
+    ? RegExp | Value
+    : Value extends number | boolean | null
+    ? Value
+    : never
+
+/**
+ * Removes all non-object members from Value
+ */
+type ObjectMembers<Value> = Value extends object ? Value : never
+
+/**
+ * The main pattern type. What a valid pattern for a value is depends on the
+ * value:
+ * - For an array, the pattern has to be a function
+ * - For an object, the pattern can be an object pattern, a wrapper pattern or a
+ *   pattern function
+ * - For a primitive value, the pattern can be a value of the same type, a
+ *   wrapper pattern or a pattern function. For strings the pattern can also be
+ *   a regular expression.
+ */
+export type PatternOf<Value, Data = unknown> =
+    // A pattern function is always a valid value. The function receives the
+    // value to be matched as argument.
+    | PatternFunction<Value, Data>
+    | (// Arrays always have to matched with a pattern function
+      // The [...] around the types are necessary to avoid
+      // distributing union types.
+      [Value] extends [any[]]
+          ? never
+          :
+                | WrapperPattern<Value, Data>
+                // Note that we are not checking types here (Value extends ...). For
+                // one, union types of mixed types (e.g. number|{x: number}) makes
+                // this a bit annoying, and what we really want to do do here is
+                // _filter_ 'Value' and allow/create an ObjectPattern for the objects in
+                // the union, and PrimitivePatterns for the primtives.
+                // ObjectMembers does the object filtering and PrimitivePattern
+                // includes type checks itself.
+                | ObjectPattern<ObjectMembers<Value>, Data>
+                | PrimitivePattern<Value>)
+
+/**
+ * Utility type to prevent TS from inferring the type from a function parameter
+ * @see https://stackoverflow.com/questions/56687668
+ */
+type NoInfer<A extends any> = [A][A extends any ? 0 : never]
+
+/**
+ * Helper type to indicate to infer the values for Value, Data from other input
+ * and not from the pattern itself.
+ */
+export type PatternOfNoInfer<Value, Data> = PatternOf<NoInfer<Value>, NoInfer<Data>>
+
+/**
+ * Context that is passed to all pattern functions during a single pattern
+ * application. Currently only holds the extracted data.
+ */
+export interface MatchContext<Data> {
+    data?: Data
+}
+
+export interface Result<Data> {
+    success: boolean
+    data?: Data
+}
```

Comment:
Should `Result` be "allowed" to hold data if the match was not successful?

```suggestion
export type Result<Data> = { success: false } | { success: true; data: Data }
```

Reply:
Probably not, good point.